File: LightsOut/LinearLightsOut.m
//
//  LinearLightsOut.m
//  LightsOut
//
//  Created by Praneet Chakraborty on 4/2/18.
//  Copyright © 2018 Praneet CSSE484. All rights reserved.
//

#import "LinearLightsOut.h"

@implementation LinearLightsOut

- (id) initWithNumLights:(NSInteger) numLights {
    self = [super init];
    self.movesTaken = 0;
    self.solved = false;
    int i;
    for(i = 0; i < numLights; i++) {
        NSInteger rand = (NSInteger)(arc4random_uniform(2));
        if(rand == 0){
            lightStates[i] = false;
            continue;
        }
        lightStates[i] = true;
    }
    return self;
}

- (BOOL) pressedLightAtIndex:(NSInteger) lightIndex {
    if (!_solved) {
        _movesTaken++;
        if (lightStates[lightIndex]) {
            lightStates[lightIndex] = false;
            [self flipLights:lightIndex];
        }
        else {
            lightStates[lightIndex] = true;
            [self flipLights:lightIndex];
        }
        int i;
        for(i = 0; i < 13; i++) {
            if (lightStates[i]) {
                return false;
            }
        }
        _solved = true;
        return true;
    }
    return true;
}

- (void) flipLights:(NSInteger) index {
    if(index == 0) { //flips second light if first light is flipped
        if(lightStates[1]) {
            lightStates[1] = false;
        }
        else {
            lightStates[1] = true;
        }
    }
    else if(index == 12) { //flips second to last light if last light is flipped
        if(lightStates[11]) {
            lightStates[11] = false;
        }
        else {
            lightStates[11] = true;
        }
    }
    else { //if any middle light is selected
        if(lightStates[index+1]) { //takes care of one directly after
            lightStates[index+1] = false;
        }
        else {
            lightStates[index+1] = true;
        }
        
        if(lightStates[index-1]) { //takes care of one directly before
            lightStates[index-1] = false;
        }
        else {
            lightStates[index-1] = true;
        }
    }
}

- (BOOL) getLightStates:(NSInteger) index {
	return lightStates[index];
}

@end

File: LightsOut/LinearLightsOut.swift
//
//  LinearLightsOut.swift
//  LightsOut
//
//  Created by Praneet Chakraborty on 3/20/18.
//  Copyright © 2018 Praneet CSSE484. All rights reserved.
//

import Foundation

import UIKit
class LinearLightsOutGame: CustomStringConvertible {
    var lightStates: [Bool]
    var moves: Int
    private var solved: Bool
    
    init(numLights: Int) {
        self.moves = 0
        self.solved = false
        self.lightStates = [Bool](repeating: true, count: numLights)
        for i in 0..<numLights {
            let n = Int(arc4random_uniform(2))
            if n == 0 {
                self.lightStates[i] = true
            }
            else {
                self.lightStates[i] = false
            }
        }
    }
    
    var description: String {
        var lightString: String = ""
        for i in 0..<lightStates.count {
            if lightStates[i] {
                lightString += "1"
            }
            else {
                lightString += "0"
            }
        }
        return "Lights: \(lightString) Moves: \(moves)"
    }
    
    func pressedLightAtIndex(_ index: Int) -> Bool{
        if !solved {
            moves += 1
            if lightStates[index] { //if selected light has true/on light
                lightStates[index] = false
                flipLights(index)
            }
            else { //if selected light has false/off light
                lightStates[index] = true
                flipLights(index)
            }
            for i in 0..<lightStates.count {
                if lightStates[i] {
                    return false;
                }
            }
            solved = true
            return true;
        }
        return true
    }
    
    private func flipLights(_ index: Int) {
        if index == 0 { //flips second light if first light is flipped
            if lightStates[1] {
                lightStates[1] = false
            }
            else {
                lightStates[1] = true
            }
        }
        else if index == lightStates.count-1 { //flips second to last light if last light is flipped
            if lightStates[lightStates.count-2] {
                lightStates[lightStates.count-2] = false
            }
            else {
                lightStates[lightStates.count-2] = true
            }
        }
        else { //if any middle light is selected
            if lightStates[index+1] { //takes care of one directly after
                lightStates[index+1] = false
            }
            else {
                lightStates[index+1] = true
            }
            
            if lightStates[index-1] { //takes care of one directly before
                lightStates[index-1] = false
            }
            else {
                lightStates[index-1] = true
            }
        }
    }
}
